### PR TITLE
DROOLS-3017: [DMN Designer] Data Types - Constraints - Base dialog / Expression component

### PR DESCRIPTION
**Pull request by karreiro**
See:
- [DROOLS-3017](https://issues.jboss.org/browse/DROOLS-3017): [DMN Designer] Data Types - Constraints - Base dialog
- [DROOLS-3020](https://issues.jboss.org/browse/DROOLS-3020): [DMN Designer] Data Types - Constraints - Expression option component

---

### CODE REVIEW

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties [hunk 1]
```diff
@@ -127,12 +127,26 @@ NameIsDefaultTypeMessage.StrongMessage=The Data Type name cannot be a default ty
...
+DataTypeConstraintModalView.Header=Data Type constraints
+DataTypeConstraintModalView.Select=Select
+DataTypeConstraintModalView.Enumeration=Enumeration
+DataTypeConstraintModalView.Expression=Expression
+DataTypeConstraintModalView.Range=Range
+DataTypeConstraintModalView.ClearAll=Clear all
+DataTypeConstraintModalView.Ok=OK
+DataTypeConstraintModalView.Cancel=Cancel
+DataTypeConstraintModalView.SelectConstraintType=Select a constraint type
+DataTypeConstraintExpressionView.ProTip=Syntax tips: Dates can be expressed using a string literal and the date() built-in function. A date in the semantic domain is a sequence of numbers for the year, month, day of month. The year must be in the range [-999,999,999..999,999,999]. Where necessary, including the valuedt function, a date value is considered to be equivalent to a date time value in which the time of day is UTC midnight (00:00:00).
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties]: It is a typo. It should simple be "..value function.."

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/common/DataTypeConstrainComponent.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.workbench.common.dmn.client.editors.types.listview.constraint.common;
+
+import elemental2.dom.Element;
+
+public interface DataTypeConstrainComponent {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/common/DataTypeConstrainComponent.java]: Should be, I suspect, `DataTypeConstraintComponent` (DataTypeConstrain**t**Component)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties (hunk 1 above)
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties]: @jomarko @manstis Yeah.. this  `valuedt` didn't sound right for me. However, this word appears like this in the spec (in the 10.3.2.3 Semantics of literals and datatypes section_) and on Liz prototype, see:

screen shot 2019-01-16 at 08 53 59

prototype_

However, I can change it, if keeping just `value` is enough.

[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties]: @karreiro OK, I see (I saw the subscript _dt_ too but thought it was a link to a footnote).

Looking at the DMN Specification (Table 66: Semantics of conversion functions) it appears that _value_ and _valuedt_ etc are not actual `Date` functions... so IDK what we should show! Perhaps @etirelli can give some insight as to what we should show! For the purpose of this PR however I think showing _valuedt_ is OK (as it's lifted "as is" from the spec'.. all be it we cannot show the subscript _dt_ part).

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/DataTypeConstraintViewTest.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+        final String expectedText = "NONE";
+        final String actualText = constraintsText.textContent;
+
+        verify(constraintsText.classList).add("none");
+        assertEquals(expectedText, actualText);
+    }
+
+    @Test
+    public void testSetupTooltip() {
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/DataTypeConstraintViewTest.java]: Fixed. Thanks!

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/DataTypeConstraintViewTest.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+        doReturn(jsonObject).when(view).makeJsonObject();
+
+        view.properties();
+
+        verify(jsonObject).put(eq("container"), eq(new JSONString("body")));
+    }
+
+    @Test
+    public void testMakeJsonObject() {
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/DataTypeConstraintViewTest.java]: Fixed. Thanks!

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/range/DataTypeConstraintRangeViewTest.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.workbench.common.dmn.client.editors.types.listview.constraint.range;
+
+public class DataTypeConstraintRangeViewTest {
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/range/DataTypeConstraintRangeViewTest.java]: Fixed. Thanks!

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/common/DataTypeConstrainComponent.java (hunk 2 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/constraint/common/DataTypeConstrainComponent.java]: Fixed. Thank you, @manstis :-)

**Pull request merged**